bigquery/google/cloud/bigquery/client.py

```diff
@@ -595,29 +596,48 @@ def run_async_query(self, job_id, query,
                         udf_resources=udf_resources,
                         query_parameters=query_parameters)
 
-    def run_sync_query(self, query, udf_resources=(), query_parameters=()):
-        """Run a SQL query synchronously.
+    def query_rows(self, query, job_config=None, job_id=None, timeout=None):
+        """Start a query job and wait for the results.
+
+        See
+        https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.query
 
         :type query: str
         :param query: SQL query to be executed
 
-        :type udf_resources: tuple
-        :param udf_resources: An iterable of
-                            :class:`google.cloud.bigquery._helpers.UDFResource`
-                            (empty by default)
+        :type job_id: str
+        :param job_id: (Optional) ID to use for the query job.
 
-        :type query_parameters: tuple
-        :param query_parameters:
-            An iterable of
-            :class:`google.cloud.bigquery._helpers.AbstractQueryParameter`
-            (empty by default)
+        :type timeout: int
+        :param timeout:
+            (Optional) How long to wait for job to complete before raising a
+            :class:`TimeoutError`.
 
-        :rtype: :class:`google.cloud.bigquery.query.QueryResults`
-        :returns: a new ``QueryResults`` instance
+        :rtype: :class:`~google.api.core.page_iterator.Iterator`
+        :returns:
+            Iterator of row data :class:`tuple`s. During each page, the
+            iterator will have the ``total_rows`` attribute set, which counts
+            the total number of rows **in the result set** (this is distinct
+            from the total number of rows in the current page:
+            ``iterator.page.num_items``).
+
+        :raises: :class:`~google.cloud.exceptions.GoogleCloudError` if the job
+            failed or  :class:`TimeoutError` if the job did not complete in the
+            given timeout.
         """
-        return QueryResults(query, client=self,
-                            udf_resources=udf_resources,
-                            query_parameters=query_parameters)
+        # TODO(swast): use job_config parameter
+        if job_id is None:
```

Comment:
Done.

job-id-is-prefix isn't ideal since I know of some use cases where a user might want to generate a deterministic job ID for a load job to prevent duplicates. Instead, configure a prefix as an option when you create a Client. But, agreed that the logic should be in one place.

Reply:
Sorry, I wasn't clear. The logic will be:
- Specify job_id, we use it exactly (that handles the use case you mention).
- Omit job_id, we generate an entirely random one
- Do something else (design TBD; in Go we have the equivalent of an add_job_id_suffix bool keyword arg) to get a random string appended to what you provide.